mlir/lib/Interfaces/DataLayoutInterfaces.cpp

```diff
@@ -744,6 +821,56 @@ LogicalResult mlir::detail::verifyDataLayoutSpec(DataLayoutSpecInterface spec,
   return success();
 }
 
+LogicalResult
+mlir::detail::verifyTargetSystemSpec(TargetSystemSpecInterface spec,
+                                     Location loc) {
+  DenseMap<StringAttr, DataLayoutEntryInterface> device_desc_keys;
+  DenseSet<TargetSystemSpecInterface::DeviceID> device_ids;
+  for (const auto &entry : spec.getEntries()) {
+    TargetDeviceSpecInterface tdd_spec = entry.second;
+    // First, verify individual target device desc specs.
+    if (failed(tdd_spec.verifyEntry(loc)))
+      return failure();
+
+    // Check that device IDs are unique across all entries.
+    TargetSystemSpecInterface::DeviceID device_id = entry.first;
+    if (!device_ids.insert(device_id).second) {
+      return failure();
+    }
+
+    // collect all the keys used by all the tdd_specs.
+    for (DataLayoutEntryInterface entry : tdd_spec.getEntries()) {
+      if (auto type = llvm::dyn_cast_if_present<Type>(entry.getKey())) {
+        // tdd_spec does not support Type as a key.
+        return failure();
+      } else {
+        device_desc_keys[entry.getKey().get<StringAttr>()] = entry;
+      }
+    }
+  }
+
+  for (const auto &kvp : device_desc_keys) {
```

Comment:
Nit: `(auto &&[first, second] : ...` works now.

Reply:
Do you need 2 `&`?